mixer/pkg/log/log.go

```diff
@@ -0,0 +1,244 @@
+//
+// The package provides direct integration with the Cobra command-line processor so make it
+// easy to build programs that use a consistent interface for logging. Here's an example
+// of a simple Cobra-based program using this log package:
+//
+//		func main() {
+//			// get the default logging options
+//			options := log.NewOptions()
+//
+//			rootCmd := &cobra.Command{
+//				Run: func(cmd *cobra.Command, args []string) {
+//
+//					// configure the logging system
+//					log.Configure(options)
+//
+//					// output some logs
+//					log.Info("Hello")
+//					log.Sync()
+//				},
+//			}
+//
+//			// add logging-specific flags to the cobra command
+//			options.AttachCobraFlags(rootCmd)
+//			rootCmd.SetArgs(os.Args[1:])
+//			rootCmd.Execute()
+//		}
+//
+// Once configured, this package intercepts the output of the standard golang "log" package as well as anything
+// sent to the global zap logger (zap.L()).
+package log
+
+import (
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// global variables against which all our logging occurs
+var logger *zap.Logger = zap.NewNop()
+var sugar *zap.SugaredLogger = logger.Sugar()
+
+// Configure initializes Istio's logging subsystem.
+//
+// You typically call this once at process startup.
+// Once this call returns, the logging system is ready to accept data.
+func Configure(options *Options) error {
+	zapConfig := zap.Config{
+		Level:       zap.NewAtomicLevelAt(zapcore.DebugLevel), // enable everything since we filter ourselves
+		Development: false,
+
+		Sampling: &zap.SamplingConfig{
+			Initial:    100,
+			Thereafter: 100,
+		},
+
+		Encoding: "console",
+		EncoderConfig: zapcore.EncoderConfig{
+			TimeKey:        "time",
+			LevelKey:       "level",
+			NameKey:        "logger",
+			CallerKey:      "caller",
+			MessageKey:     "msg",
+			StacktraceKey:  "stack",
+			LineEnding:     zapcore.DefaultLineEnding,
+			EncodeLevel:    zapcore.LowercaseLevelEncoder,
+			EncodeCaller:   zapcore.ShortCallerEncoder,
+			EncodeTime:     zapcore.ISO8601TimeEncoder,
+			EncodeDuration: zapcore.StringDurationEncoder,
+		},
+
+		OutputPaths:       options.OutputPaths,
+		ErrorOutputPaths:  []string{"stderr"},
+		DisableCaller:     true,
+		DisableStacktrace: true,
+	}
+
+	if options.JSONEncoding {
+		zapConfig.Encoding = "json"
+	}
+
+	var err error
+	logger, err = zapConfig.Build(zap.WrapCore(func(c zapcore.Core) zapcore.Core {
+		return &wrapper{
+			original: c,
+			options:  options,
+		}
+	}))
+
+	if err != nil {
+		return err
+	}
+
+	sugar = logger.Sugar()
+
+	// capture global zap logging and force it through our logger
+	_ = zap.ReplaceGlobals(logger)
+
+	// capture standard golang "log" package output and force it through our logger
+	_ = zap.RedirectStdLog(logger)
+
+	return nil
+}
+
+// Debug outputs a message at debug level.
+// This call is a wrapper around [Logger.Debug](https://godoc.org/go.uber.org/zap#Logger.Debug)
+func Debug(msg string, fields ...zapcore.Field) {
```

Comment:
This should also be an interface. That way we can pass it to other entities like adapters.

Reply:
what about an approach around putting a (named) logger into the context passed to adapters (and the like)?  with the new `L()` and `S()` and `With()` funcs, there should be a way to pass loggers into adapters. That isn't quite the same as an interface, but might achieve the same purpose.

Reply:
I think the implementation of `env.Logger` should be backed by this logger, and I'm not sure we need more than that. I'm not really a fan of sticking the logger into the context, I think it'll result in a lot more boiler plate (you always have to check if you actually have it in hand, then what do you do if you don't get a logger? wrap every log site in an if? ew).

Reply:
fwiw, extracting the logger from the context is the approach taken by the grpc ecosystem libs. you end up with code like: `logger := grpc_zap.Extract(ctx)`. It isn't all that ugly.